fvm/evm/emulator/database/database.go

```diff
@@ -0,0 +1,341 @@
+	// these type of locking if the underlying structure (atree)
+	// has such protections or if EVM really needs it
+	lock sync.RWMutex
+}
+
+// NewDatabase returns a wrapped map that implements all the required database interface methods.
+func NewDatabase(led atree.Ledger, flowEVMRootAddress flow.Address) (*Database, error) {
+	baseStorage := atree.NewLedgerBaseStorage(led)
+
+	storage, err := NewPersistentSlabStorage(baseStorage)
+	if err != nil {
+		return nil, handleError(err)
+	}
+
+	db := &Database{
+		led:                led,
+		flowEVMRootAddress: flowEVMRootAddress,
+		storage:            storage,
+	}
+
+	err = db.retrieveOrCreateMapRoot()
+	if err != nil {
+		return nil, err
+	}
+	return db, nil
+}
+
+func (db *Database) retrieveOrCreateMapRoot() error {
+	rootIDBytes, err := db.led.GetValue(db.flowEVMRootAddress.Bytes(), []byte(FlowEVMRootSlabKey))
+	if err != nil {
+		return handleError(err)
+	}
+
+	var m *atree.OrderedMap
+	if len(rootIDBytes) == 0 {
+		m, err = atree.NewMap(db.storage, atree.Address(db.flowEVMRootAddress), atree.NewDefaultDigesterBuilder(), typeInfo{})
+		if err != nil {
+			return handleError(err)
+		}
+	} else {
+		storageID, err := atree.NewStorageIDFromRawBytes(rootIDBytes)
+		if err != nil {
+			return handleError(err)
+		}
+		m, err = atree.NewMapWithRootID(db.storage, storageID, atree.NewDefaultDigesterBuilder())
+		if err != nil {
+			return handleError(err)
+		}
+	}
+	db.atreemap = m
+	return nil
+}
+
+func (db *Database) storeMapRoot() error {
+	rootIDBytes := make([]byte, StorageIDSize)
+	_, err := db.atreemap.StorageID().ToRawBytes(rootIDBytes)
+	if err != nil {
+		return handleError(err)
+	}
+	err = db.led.SetValue(db.flowEVMRootAddress.Bytes(), []byte(FlowEVMRootSlabKey), rootIDBytes[:])
+	if err != nil {
+		return handleError(err)
+	}
+	return nil
+}
+
+// Get retrieves the given key if it's present in the key-value store.
+func (db *Database) Get(key []byte) ([]byte, error) {
+	db.lock.RLock()
+	defer db.lock.RUnlock()
+
+	return db.get(key)
+}
+
+func (db *Database) get(key []byte) ([]byte, error) {
+	data, err := db.atreemap.Get(compare, hashInputProvider, NewByteStringValue(key))
+	if err != nil {
+		return nil, handleError(err)
+	}
+
+	v, err := data.StoredValue(db.atreemap.Storage)
+	if err != nil {
+		return nil, handleError(err)
+	}
+
+	return v.(ByteStringValue).Bytes(), nil
+}
+
+// Put inserts the given value into the key-value store.
+func (db *Database) Put(key []byte, value []byte) error {
+	db.lock.Lock()
+	defer db.lock.Unlock()
+	return db.put(key, value)
+}
+
+func (db *Database) put(key []byte, value []byte) error {
+	_, err := db.atreemap.Set(compare, hashInputProvider, NewByteStringValue(key), NewByteStringValue(value))
+	return handleError(err)
+}
+
+// Has checks if a key is present in the key-value store.
+func (db *Database) Has(key []byte) (bool, error) {
+	db.lock.RLock()
+	defer db.lock.RUnlock()
+
+	data, err := db.get(key)
+	return len(data) > 0, err
```

Comment:
Probably should use `OrderedMap.Has` directly because `OrderedMap.Get` returns error if key doesn't exist.

```suggestion
	exist, err := db.atreemap.Has(compare, hashInputProvider, key)
	if err != nil {
		return false, handleError(err)
	}
	return exist, nil
```